source/reduce/reduction_pass.cpp

```diff
@@ -0,0 +1,77 @@
+// Copyright (c) 2018 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <algorithm>
+
+#include "reduction_pass.h"
+
+#include "source/opt/build_module.h"
+
+namespace spvtools {
+namespace reduce {
+
+std::vector<uint32_t> ReductionPass::ApplyReduction(
+    const std::vector<uint32_t>& binary) {
+  std::unique_ptr<opt::IRContext> context =
+      BuildModule(target_env_, consumer_, binary.data(), binary.size());
+  assert(context);
+
+  std::vector<std::unique_ptr<ReductionOpportunity>> opportunities =
+      GetAvailableOpportunities(context.get());
+
+  if (!is_initialized_) {
+    assert(granularity_ == 0);
+    assert(index_ == 0);
+    is_initialized_ = true;
+    index_ = 0;
+    granularity_ = std::max((uint32_t)1, (uint32_t)opportunities.size());
```

Comment:
If you remove the `assert(granularity_ > 0)` below (or move after the subsequent if) you can drop this max. It appears you would still always get a minimum granularity of 1.

Reply:
Agreed, done.